### PR TITLE
feat(users): user functionality

### PR DESCRIPTION
**Pull request by Ferkalek**
feat(users): added users functionality

### CODE REVIEW

**Comment by ArtemLunt** [comment]
General: PR name should be in the same format as commit message.

File: src/app/actions/user.actions.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { IUserDTO } from '../users/users.interface';
+
+export class AddUserAction {
+    static readonly type = '[User API] Add User';
+
+    constructor(public payload: IUserDTO) { }
```
Comment (ArtemLunt) [inline src/app/actions/user.actions.ts:6]: Why payload?
I guess it should be `user`.
Also we can make it readonly.

File: src/app/app.module.ts [hunk 2]
```diff
@@ -33,7 +34,9 @@ import { NotFoundPageComponent } from './not-found-page/not-found-page.component
     MatToolbarModule,
     MatButtonModule,
 
-    NgxsModule.forRoot(),
+    NgxsModule.forRoot([
```
Comment (ArtemLunt) [inline src/app/app.module.ts:37]: We should load this state in user feature module.

File: src/app/state/user.state.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    users: IUserDTO[];
+}
+
+@State<UsersStateModel>({
+    name: 'users',
+    defaults: {
+        users: []
+    }
+})
+// @Injectable() // ???
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:15]: What a comments?

File: src/app/models/tutorial.model.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export interface Tutorial {
```
Comment (ArtemLunt) [inline src/app/models/tutorial.model.ts:1]: `ITutorial` pls

File: src/app/state/user.state.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+export class UsersState {
+
+    @Selector()
+    static getUsers(state: UsersStateModel) {
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:19]: Usually selectors are doesn't have to be getters.
Just `users`.

File: src/app/actions/user.actions.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { IUserDTO } from '../users/users.interface';
+
+export class AddUserAction {
```
Comment (ArtemLunt) [inline src/app/actions/user.actions.ts:3]: There should be multiple users, so User**s**

File: src/app/state/user.state.ts (hunk 5 above)
Comment (ArtemLunt) [reply src/app/state/user.state.ts:19]: Also you can make it prettier using destructuring.

File: src/app/state/user.state.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        return state.users;
+    }
+
+    @Action(AddUserAction)
+    add({getState, patchState }: StateContext<UsersStateModel>, { payload }:AddUserAction) {
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:24]: `addUsers`

File: src/app/state/user.state.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        const state = getState();
+        patchState({
+            users: [...state.users, payload]
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:27]: Pls use [state operator append](https://www.ngxs.io/advanced/operators#example) here.

File: src/app/users/users-list/users-list.component.scss [hunk 9]
```diff
@@ -0,0 +1,19 @@
+.user-list {
+    padding: 0 15px 50px;
+}
+
+.user-card {
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.scss:5]: I guess it should be inside of `.users-list` selector to keep the stucture.

File: src/app/users/users-list/users-list.component.ts [hunk 10]
```diff
@@ -1,9 +1,34 @@
...
 
 @Component({
   selector: 'app-users-list',
   templateUrl: './users-list.component.html',
   styleUrls: ['./users-list.component.scss'],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class UsersListComponent {}
+export class UsersListComponent {
+  users$: Observable<IUserDTO[]>;
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:15]: readonly!

File: src/app/users/users-list/users-list.component.ts [hunk 11]
```diff
@@ -1,9 +1,34 @@
...
+  
+  constructor(
+    private _usersService: UsersService,
+    private store: Store
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:19]: prefix!

File: src/app/users/users-list/users-list.component.ts [hunk 12]
```diff
@@ -1,9 +1,34 @@
...
+  ) {
+    this.users$ = this.store.select(state => state.users.users)
+  }
+
+  ngOnInit(): void {
+      this._usersService.getUsers()
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:25]: Unsubscribe!

File: src/app/users/users-list/users-list.component.ts [hunk 13]
```diff
@@ -1,9 +1,34 @@
...
+        .subscribe(users => {
+          users.forEach(user => this.addUser(user))
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:27]: Why you're adding users one by one?

File: src/app/users/users-list/users-list.component.ts (hunk 10 above)
Comment (ArtemLunt) [reply src/app/users/users-list/users-list.component.ts:15]: We're usually using decorator `@Select()`.

File: src/app/users/users-list/users-list.component.ts [hunk 14]
```diff
@@ -1,9 +1,34 @@
...
+        })
+  }
+
+  addUser(user: IUserDTO): void {
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:31]: Why do we need this?

File: src/app/actions/user.actions.ts (hunk 1 above)
Comment (ArtemLunt) [reply src/app/actions/user.actions.ts:6]: Also pls take an array here.

File: src/app/users/users-repository.service.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { IUsersDTO } from './users.interface';
+
+const BASE_URL = 'https://reqres.in/api/users';
```
Comment (ArtemLunt) [inline src/app/users/users-repository.service.ts:6]: What is that?
It shouldn't be here at all.

File: src/app/users/users-repository.service.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+@Injectable({
+    providedIn: 'root',
+})
+export class UsersRepositoryService {
+    constructor(
+        private _http: HttpClient
+    ) { }
+
+    sendGetUsersRequest(): Observable<IUsersDTO> {
+        return this._http.get<IUsersDTO>(`${BASE_URL}?page=2`);
```
Comment (ArtemLunt) [inline src/app/users/users-repository.service.ts:17]: Pls use `HttpParams` for get params.

Comment (ArtemLunt) [reply src/app/users/users-repository.service.ts:17]: `page` param should be in enum.

File: src/app/users/users.interface.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
+export interface IUserDTO {
+    id: number;
+    email: string;
+    first_name: string;
+    last_name: string;
+    avatar: string;
+}
+
+interface IAdDTO {
```
Comment (ArtemLunt) [inline src/app/users/users.interface.ts:9]: ???

File: src/app/users/users.interface.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    company: string;
+    url: string;
+    text: string;
+}
+
+export interface IUsersDTO {
```
Comment (ArtemLunt) [inline src/app/users/users.interface.ts:15]: There should be separate `IPage` interface with a pagination properties and separate interface for Users page.

File: src/app/users/users-repository.service.ts (hunk 16 above)
Comment (ArtemLunt) [reply src/app/users/users-repository.service.ts:17]: Also, leave todo for pagination here pls.

File: src/app/users/users.interface.ts (hunk 18 above)
Comment (ArtemLunt) [reply src/app/users/users.interface.ts:15]: Also, pls add alias for this generic interface, smth like `interface IUserPage extends IPage<IUserDTO> {}`

File: src/app/users/users.service.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+})
+export class UsersService {
+    constructor(
+        private _usersRepositoryService: UsersRepositoryService
+    ) { }
+
+    getUsers(): Observable<IUserDTO[]> {
+        return this._usersRepositoryService.sendGetUsersRequest()
+            .pipe(
+                map(dataUsers => dataUsers.data)
```
Comment (ArtemLunt) [inline src/app/users/users.service.ts:19]: `({data}) => data` would be more pretty-shine

File: src/styles.scss [hunk 20]
```diff
@@ -7,3 +7,21 @@ h1, h2 {
...
   margin: 1em 0 1.5em;
 }
+//-------------
+body, html {
+  margin: 0;
+  padding: 0;
+  font-family: 'Arial';
+}
+
+.left, .right {
```
Comment (ArtemLunt) [inline src/styles.scss:17]: Pls rid of that

File: src/styles.scss [hunk 21]
```diff
@@ -7,3 +7,21 @@ h1, h2 {
   text-align: center;
   margin: 1em 0 1.5em;
 }
+//-------------
```
Comment (ArtemLunt) [inline src/styles.scss:10]: ???

File: src/app/app.module.ts (hunk 2 above)
Comment (Ferkalek) [reply src/app/app.module.ts]: done

File: src/app/state/user.state.ts (hunk 3 above)
Comment (Ferkalek) [reply src/app/state/user.state.ts]: changed

File: src/app/models/tutorial.model.ts (hunk 4 above)
Comment (Ferkalek) [reply src/app/models/tutorial.model.ts]: deleted

File: src/app/state/user.state.ts (hunk 5 above)
Comment (Ferkalek) [reply src/app/state/user.state.ts]: changed

File: src/app/users/users.interface.ts (hunk 17 above)
Comment (Ferkalek) [reply src/app/users/users.interface.ts:9]: discussed

File: src/app/users/users.interface.ts (hunk 18 above)
Comment (Ferkalek) [reply src/app/users/users.interface.ts]: changed

File: src/app/users/users.service.ts (hunk 19 above)
Comment (Ferkalek) [reply src/app/users/users.service.ts]: thanks, changed

File: src/styles.scss (hunk 20 above)
Comment (Ferkalek) [reply src/styles.scss]: it's useless
deleted

File: src/styles.scss (hunk 21 above)
Comment (Ferkalek) [reply src/styles.scss]: deleted

File: src/app/actions/user.actions.ts (hunk 1 above)
Comment (Ferkalek) [reply src/app/actions/user.actions.ts]: not sure that I understood first part.
changed to load all users in one time

File: src/app/state/user.state.ts (hunk 7 above)
Comment (Ferkalek) [reply src/app/state/user.state.ts]: done

File: src/app/state/user.state.ts (hunk 8 above)
Comment (Ferkalek) [reply src/app/state/user.state.ts]: tried but didn't manage it after changed on all loading users

File: src/app/users/users-list/users-list.component.scss (hunk 9 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.scss]: changed

File: src/app/users/users-list/users-list.component.ts (hunk 10 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.ts:16]: changed but have a question

File: src/app/users/users-list/users-list.component.ts (hunk 11 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.ts]: done

File: src/app/users/users-list/users-list.component.ts (hunk 12 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.ts:27]: added our abstartact class

File: src/app/users/users-list/users-list.component.ts (hunk 13 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.ts]: changed

File: src/app/users/users-list/users-list.component.ts (hunk 14 above)
Comment (Ferkalek) [reply src/app/users/users-list/users-list.component.ts]: deleted

File: src/app/users/users-repository.service.ts (hunk 15 above)
Comment (Ferkalek) [reply src/app/users/users-repository.service.ts]: changed but have a question

File: src/app/users/users-repository.service.ts (hunk 16 above)
Comment (Ferkalek) [reply src/app/users/users-repository.service.ts]: HttpParams  added but it is not working

File: src/app/state/user.state.ts [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    defaults: {
+        users: []
+    }
+})
+@Injectable()
+export class UsersState {
+
+    @Selector()
+    static users(state: UsersStateModel) {
+        return state.users;
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:21]: pls use destructuring

File: src/app/state/user.state.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Action(AddUserAction)
+    addUsers({getState, patchState }: StateContext<UsersStateModel>, { payload }: AddUserAction) {
+        const state = getState();
+        patchState({
+            users: [...state.users, ...payload]
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:28]: State operators pls.

File: src/app/users/users-list/users-list.component.ts [hunk 24]
```diff
@@ -1,9 +1,31 @@
...
 @Component({
   selector: 'app-users-list',
   templateUrl: './users-list.component.html',
   styleUrls: ['./users-list.component.scss'],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class UsersListComponent {}
+export class UsersListComponent extends ASubscriptionCollector {
+  // readonly users$: Observable<IUserDTO[]>; // ???
+  @Select(state => state.users.users) users$: Observable<any>;
```
Comment (ArtemLunt) [inline src/app/users/users-list/users-list.component.ts:17]: readonly as well.

File: src/app/state/user.state.ts [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+@State<UsersStateModel>({
+    name: 'users',
+    defaults: {
+        users: []
+    }
+})
+@Injectable()
+export class UsersState {
+
```
Comment (ArtemLunt) [inline src/app/state/user.state.ts:18]: Extra breakline

**Pull request merged**